quantumblacklabs/kedro-viz
# Add data layers

## Description

- Separate graph nodes vertically (by layer)
- Add layer bands, and highlight them on hover

![image](https://user-images.githubusercontent.com/1155816/78349550-aeb99500-759b-11ea-8865-a486fca42c44.png)

## Development notes

This is a really big PR - sorry about that!

The layered rankings depends on being able to set `ranker: "none"`, a feature in Dagre that currently only exists on [a PR that has been open since June last year](https://github.com/dagrejs/dagre/pull/271). To get this to work we'll either need to fork Dagre or see if we can get the PR resurrected and merged. Else, we'll need to switch to a new DAG layout engine or write our own. In the short term, I've mirrored the branch from the PR on my own GitHub account and linked to it in package.json, so that we at least have ownership of it and can be sure that it won't be deleted or altered.

### Tasks

- [x] Prototype layer zoom/pan on background strips and label layers
- [x]  Limit min/max zoom scale extent
- [x] Limit min/max zoom translate extent
- [x]  Use manual ranking in Dagre
- [ ] Get manual ranking merged into Dagre
- [x] Calculate manual ranks on the client side
- [x] Demonstrate switching between rankers in Dagre
- [x] Remove temporary ranker switch toggle
- [x] Fix layer label overlaps
- [x] Fix transition delay bug
- [x] Add layer toggle button

## QA notes

There's a new test dataset called 'layers.mock.js'. To see layers in action, add `?data=layers` to the URL.

You can toggle them on/off with this button:
![image](https://user-images.githubusercontent.com/1155816/78351599-fbeb3600-759e-11ea-8988-15454301ee22.png)

You can see the layer bands by hovering over them. Note that they're not shown when hovering over a node - @GabrielComymQB is this okay?

## Checklist


## Legal notice


- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
